litt/disktable/segment/segment.go

```diff
@@ -0,0 +1,734 @@
+	shards := make([]*valueFile, metadata.shardingFactor)
+	for shard := uint32(0); shard < metadata.shardingFactor; shard++ {
+		values, err := loadValueFile(logger, index, shard, parentDirectories)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open value file: %v", err)
+		}
+		shards[shard] = values
+	}
+
+	segment := &Segment{
+		logger:            logger,
+		fatalErrorHandler: fatalErrorHandler,
+		index:             index,
+		metadata:          metadata,
+		keys:              keys,
+		shards:            shards,
+		keyFileSize:       keyFileSize,
+		deletionChannel:   make(chan struct{}, 1),
+	}
+
+	// Segments are returned with an initial reference count of 1, as the caller of the constructor is considered to
+	// have a reference to the segment.
+	segment.reservationCount.Store(1)
+
+	return segment, nil
+
+}
+
+// Size returns the size of the segment in bytes. Counts bytes that are on disk or that will eventually end up on disk.
+// This method is not thread safe, and should not be called concurrently with methods that modify the segment.
+func (s *Segment) Size() uint64 {
+	size := s.metadata.Size()
+
+	if s.IsSealed() {
+		// This segment is immutable, so it's thread safe to query the files directly.
+		size += s.keys.Size()
+		for _, shard := range s.shards {
+			size += shard.Size()
+		}
+	} else {
+		// This segment is mutable. We must use our local reckoning of the sizes of the files.
+		size += s.keyFileSize
+		for _, shardSize := range s.shardSizes {
+			size += shardSize
+		}
+	}
+
+	return size
+}
+
+// lookForFile looks for a file in a list of directories. It returns an error if the file appears
+// in more than one directory, and an empty string if the file is not found. If the file is found and
+// there are no errors, this method returns the path to the file.
+func lookForFile(directories []string, fileName string) (string, error) {
+	locations := make([]string, 0, 1)
+	for _, directory := range directories {
+		potentialLocation := path.Join(directory, fileName)
+		exists, err := util.Exists(potentialLocation)
+		if err != nil {
+			return "", fmt.Errorf("failed to check if file %s exists: %v", potentialLocation, err)
+		}
+		if exists {
+			locations = append(locations, potentialLocation)
+		}
+	}
+
+	if len(locations) > 1 {
+		return "", fmt.Errorf("file %s found in multiple directories: %v", fileName, locations)
+	}
+
+	if len(locations) == 0 {
+		return "", nil
+	}
+	return locations[0], nil
+}
+
+// SetNextSegment sets the next segment in the chain.
+func (s *Segment) SetNextSegment(nextSegment *Segment) {
+	nextSegment.Reserve()
+	s.nextSegment = nextSegment
+}
+
+// GetShard returns the shard number for a key.
+func (s *Segment) GetShard(key []byte) uint32 {
+	if s.metadata.shardingFactor == 1 {
+		// Shortcut: if we have one shard, we don't need to hash the key to figure out the mapping.
+		return 0
+	}
+
+	return util.HashKey(key, s.metadata.salt) % s.metadata.shardingFactor
```

Comment:
how does this work for dynamic sharding if that's needed?

Reply:
When the sharding factor changes, the current segment is sealed and a new segment is created with the new sharding factor. Old segments keep the sharding factor that they originally had, since changing the sharding factor for existing segments would require a very expensive migration.

Reply:
So the history sharding factor is kept and old segments are separately handled for reading, if there is new sharding factor? (didn't read into that details)

Reply:
Yes. Each segment has a metadata file that contains a variety of data, including the sharding factor that was used to write the segment. When reading the segment, littDB uses that sharding factor.

There is a method on `litt.Table` that allows you to modify the sharding factor. This method is allowed to be called at any time.

```
	// SetShardingFactor sets the number of write shards used. Increasing this value increases the number of parallel
	// writes that can be performed.
	SetShardingFactor(shardingFactor uint32) error
```